### PR TITLE
Replace flake8, bugbear, and isort with ruff

### PR DESCRIPTION
**Pull request by cclauss**
* [ ] I have added an entry to `docs/changelog.md`

## Summary of changes

[Ruff](https://beta.ruff.rs) supports [over 500 lint rules](https://beta.ruff.rs/docs/rules) including flake8, flake8-bugbear, isort, pylint, and pyupgrade and is written in Rust for speed.
Also added rules for pylint conventions, pylint errors, and pylint warnings.

## Test plan

Tested by running
```
ruff .
pre-commit run --all-files
```
Closes #945

### CODE REVIEW

File: docs/programs-to-try.md [hunk 1]
```diff
@@ -58,12 +58,12 @@ easily create and publish new Python packages
 pipx run create-python-package
 ```
 
-### flake8
```
Comment (uranusjr) [inline docs/programs-to-try.md:61]: We should keep flake8 here (and add ruff separately), this is just “programs to try” and having flake8 here is correct.

**Comment by uranusjr** [comment]
Personally I’m fine with this, but other maintainers may have different opinions.

**Comment by meowmeowmeowcat** [comment]
I am +1 on this. 👍

File: docs/programs-to-try.md (hunk 1 above)
Comment (cclauss) [reply docs/programs-to-try.md:61]: FIxed.

File: docs/programs-to-try.md [hunk 2]
```diff
@@ -59,7 +59,6 @@ pipx run create-python-package
 ```
 
 ### flake8
```
Comment (cclauss) [inline docs/programs-to-try.md:61]:
```suggestion
### flake8

```

**Review by uranusjr — APPROVED** [review]

**Pull request merged**